R/zchunk_L102.GDP.R

```diff
@@ -1,39 +1,118 @@
+        future$scenario <- 'scen'
+        drop.scenario <- TRUE
+    }
+    else {
+        drop.scenario <- FALSE
+    }
+
+    ## Find the base year
+    base.year <- max(dplyr::intersect(past$year, future$year))
+    ## Base year gdp from the future dataset
+    baseyear.future.gdp <- filter(future, year==base.year) %>%
+        rename(base.gdp = gdp) %>%
+        select(-year)
+
+    gdp.future.ratio <- filter(future, year > base.year) %>%
+        left_join_error_no_match(baseyear.future.gdp, by=c('scenario',grouping)) %>%
+        mutate(gdp.ratio = gdp / base.gdp) %>%
+        select(one_of(c('scenario', grouping, 'year', 'gdp.ratio')))
+
+    ## add the scenario column to the past
+    gdp.past <- tidyr::crossing(past, scenario = unique(gdp.future.ratio[['scenario']]))
+    baseyear.past.gdp <- filter(gdp.past, year == base.year) %>%
+        rename(base.gdp = gdp) %>%
+        select(-year)
+
+    rslt <- left_join(baseyear.past.gdp, gdp.future.ratio,
+                      by=c('scenario',grouping)) %>%
+        mutate(gdp = base.gdp * gdp.ratio) %>%
+        select(one_of(c('scenario', grouping, 'year', 'gdp'))) %>%
+        bind_rows(gdp.past, .)
+
+    if(drop.scenario)
+        select(rslt, -scenario)
+    else
+        rslt
+}
+
 #' module_socioeconomics_L102.GDP
 #'
-#' Briefly describe what this chunk does.
+#' Prepare historical and future GDP time series.  On the historical side, this
+#' amounts to aggregating country-level GDP to GCAM regions.  On the future side
+#' we create a time series for each of a variety of future scenarios.  The
+#' outputs include GDP, pcGDP, and the PPP-MER conversion factor, all tabulated
+#' by GCAM region.
+#'
+#' The scenarios generated include the SSPs, the gSSPs (SSPs modified by
+#' near-term IMF projections), and GCAM3 legacy scenario (unused) to be used for replicating old scenarios. 
+#' GDP outputs are in millions of 1990 USD, Market Exchange Rate (measured in 2010) is used for foreign currency. 
+#' Per-capita values are in thousands of 1990 USD.
+#'
 #'
 #' @param command API command to execute
 #' @param ... other optional parameters, depending on command
 #' @return Depends on \code{command}: either a vector of required inputs,
 #' a vector of output names, or (if \code{command} is "MAKE") all
-#' the generated outputs: \code{L102.gdp_mil90usd_Scen_R_Y}, \code{L102.pcgdp_thous90USD_Scen_R_Y}, \code{L102.gdp_mil90usd_GCAM3_R_Y}, \code{L102.gdp_mil90usd_GCAM3_ctry_Y}, \code{L102.pcgdp_thous90USD_GCAM3_R_Y}, \code{L102.pcgdp_thous90USD_GCAM3_ctry_Y}, \code{L102.pcgdp_thous90USD_GCAM3_ctry_Y}, \code{L102.PPP_MER_R}. The corresponding file in the
+#' the generated outputs: \code{L102.gdp_mil90usd_Scen_R_Y}, \code{L102.pcgdp_thous90USD_Scen_R_Y}, \code{L102.gdp_mil90usd_GCAM3_R_Y}, \code{L102.gdp_mil90usd_GCAM3_ctry_Y}, \code{L102.pcgdp_thous90USD_GCAM3_R_Y}, \code{L102.pcgdp_thous90USD_GCAM3_ctry_Y}, \code{L102.PPP_MER_R}. The corresponding file in the
```

Comment:
To that I'll add, if someone thinks they might need to use these some day, they are welcome to add them back.  I would recommend doing them in a separate chunk, using the outputs of this chunk inputs to the new chunk where required.

I would also recommend that the work to reinstate the GCAM3 scenarios be charged to the project that actually requires them because everybody I've talked to seems to think we have no use for them in the SFA.